File: java/Topicos avanzados de programacion/Calculadora/src/eventoPulsado.java
import javax.swing.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.ArrayList;
import java.util.Objects;

public class eventoPulsado {
    String cadena;
    String cadenaRes;//regresa un resultado
    private static ArrayList<String> conjuntoEventosM=new ArrayList<>();
    private static Boolean punto=true;
     eventoOperacionInterna operacion=new eventoOperacionInterna();



    eventoPulsado(String cadena, String cadenaRes){
        this.cadena = cadena;
        this.cadenaRes = cadenaRes;



    }

    public String getCadena() {
        return cadena;
    }

    public String getCadenaRes() {
        return cadenaRes;
    }

    public void setCadenaRes(String cadenaRes) {
        this.cadenaRes = cadenaRes;
    }


    public void setCadena(String cadena) {
        this.cadena = cadena;
    }

    public void action() {
        String apuntador="";



    if(cadenaRes!=null&& !cadenaRes.isEmpty()) apuntador=String.valueOf(cadenaRes.charAt(cadenaRes.length()-1));
     //   System.out.println("");
      //  System.out.println(apuntador);


        System.out.println(punto);
        switch(cadena){
            case "+": eventoSuma(apuntador); break;
            case "-": eventoResta(apuntador); break;
            case "x": eventoMultiplicacion(apuntador); break;
            case ".": eventoPunto(apuntador); break;
            case "C": eventoC(); break;
            case "/": eventoDiagonal(apuntador); break;
            case "MR": eventoMR(); break;
            case "ML": eventoML(); break;
            case "X^n": eventoXn(apuntador); break;
            case "√": eventoRaiz(apuntador); break;
            case "M+": eventoMSuma(apuntador); break;
            case "M-": eventoMResta(apuntador); break;
            case "=": eventoIgual(); break;

            default :
                cadenaRes=cadenaRes+cadena;
                break;
        }
        if (Objects.equals(cadena, ".")) punto=false;
    }

    public void eventoSuma(String apuntador){
        eventoSignos(apuntador);

    }
    public void eventoResta(String apuntador){
        eventoSignos(apuntador);
    }
    public void eventoMultiplicacion(String apuntador){
        eventoSignos(apuntador);
    }
    public void eventoPunto(String apuntador){
        if (punto) eventoSignos(apuntador);
    }
    public void eventoC(){

        punto=true;
        cadenaRes="";
    }

    public void eventoDiagonal(String apuntador){
        eventoSignos(apuntador);
    }
    public void eventoMR(){
      //  String cadenaTemporal= operacion.operation(cadenaRes);
        String cadenaTemporal="";
        for (String evento: conjuntoEventosM){
            //System.out.printf(cadenaTemporal);
            cadenaTemporal+=evento;
        }
        /*  if (cadenaTemporal.isEmpty())  JOptionPane.showMessageDialog(null,"Datos mal implementados");
        else JOptionPane.showMessageDialog(null,cadenaTemporal+"\nResultado="+resultadoM);*/
        cadenaRes= operacion.operation(cadenaTemporal);
      //  conjuntoEventosM.clear();
    }
    public void eventoML(){
        if (eventosM(cadenaRes)) JOptionPane.showMessageDialog(null,operacion.factorial(cadenaRes));

    }
    public void eventoXn(String apuntador){

        if (eventosM(apuntador)) cadenaRes=cadenaRes+"^";


    }
    public void eventoRaiz(String apuntador){
        //√
        if (!Objects.equals(apuntador, "+") && !Objects.equals(apuntador, "-")
                && !Objects.equals(apuntador, "x") && !Objects.equals(apuntador, "/")&&
                !Objects.equals(apuntador, ".")&& !Objects.equals(apuntador, "^")&& !Objects.equals(apuntador, "√"))
            cadenaRes=cadenaRes+"√";



    }
    public void eventoMSuma(String apuntador){
        String cadenaTemporal="";
        if (eventosM(apuntador)) {
            if (conjuntoEventosM.isEmpty()) cadenaTemporal=operacion.operation(cadenaRes);
            else cadenaTemporal="+"+operacion.operation(cadenaRes);

           // System.out.printf(cadenaTemporal);
            conjuntoEventosM.add(cadenaTemporal);
        }

    }
    public void eventoMResta(String apuntador){
        String cadenaTemporal="";
        if (eventosM(apuntador)) {

            cadenaTemporal="-"+operacion.operation(cadenaRes);
           // System.out.printf(cadenaTemporal);
            conjuntoEventosM.add(cadenaTemporal);
        }
    }
    public void eventoIgual(){
        cadenaRes= operacion.operation(cadenaRes);
      //  punto=true;



    }


public void eventoSignos(String apuntador){



    if (!cadenaRes.isEmpty() &&!Objects.equals(apuntador, "+") && !Objects.equals(apuntador, "-")
            && !Objects.equals(apuntador, "x") && !Objects.equals(apuntador, "/")&&
            !Objects.equals(apuntador, ".")&& !Objects.equals(apuntador, "^")&& !Objects.equals(apuntador, "√")) cadenaRes=cadenaRes+cadena;
}

public boolean eventosM(String apuntador){
    if (!cadenaRes.isEmpty() &&!Objects.equals(apuntador, "+") && !Objects.equals(apuntador, "-")
            && !Objects.equals(apuntador, "x") && !Objects.equals(apuntador, "/")&&
            !Objects.equals(apuntador, ".")&& !Objects.equals(apuntador, "^")&& !Objects.equals(apuntador, "√"))
        return true;
else return false;
    }
}

File: C#/SharpDeveloper/GUI_CALCULADORA/GUI_CALCULADORA/eventoPulsado.cs
/*
 * Created by SharpDevelop.
 * User: Alexis
 * Date: 10/02/2025
 * Time: 12:44 a. m.
 * 
 * To change this template use Tools | Options | Coding | Edit Standard Headers.
 */
//using System;

//namespace GUI_CALCULADORA
using System;
using System.Collections.Generic;

public class eventoPulsado
{
    private string cadena;
    private string cadenaRes;
    private static List<string> conjuntoEventosM = new List<string>();
    private static bool punto = true;
  eventoOperacionInterna operacion=new eventoOperacionInterna();

    public eventoPulsado(string cadena, string cadenaRes)
    {
        this.cadena = cadena;
        this.cadenaRes = cadenaRes;
    }

    public string GetCadena() {return cadena;}
    public string GetCadenaRes() {return cadenaRes;}
    public void SetCadena(string value) {cadena = value;}
    public void SetCadenaRes(string value) {cadenaRes = value;}

    public void Action()
    {
        string apuntador ="";

        if(!string.IsNullOrEmpty(cadenaRes)) apuntador = cadenaRes.Substring(cadenaRes.Length - 1, 1);
        Console.WriteLine(punto);
        
        switch (cadena)
        {
            case "+": EventoSuma(apuntador); break;
            case "-": EventoResta(apuntador); break;
            case "x": EventoMultiplicacion(apuntador); break;
            case ".": EventoPunto(apuntador); break;
            case "C": EventoC(); break;
            case "/": EventoDiagonal(apuntador); break;
            case "MR": EventoMR(); break;
            case "ML": EventoML(); break;
            case "X^n": EventoXn(apuntador); break;
            case "√": EventoRaiz(apuntador); break;
            case "M+": EventoMSuma(apuntador); break;
            case "M-": EventoMResta(apuntador); break;
            case "=": EventoIgual(); break;
            default: cadenaRes += cadena; break;
        }
        if (cadena == ".") punto = false;
    }

    private void EventoSuma(string apuntador) { EventoSignos(apuntador);}
    private void EventoResta(string apuntador) { EventoSignos(apuntador);}
    private void EventoMultiplicacion(string apuntador) { EventoSignos(apuntador);}
    private void EventoPunto(string apuntador) { if (punto) EventoSignos(apuntador); }
    private void EventoC() { punto = true; cadenaRes = ""; }
    private void EventoDiagonal(string apuntador) {EventoSignos(apuntador);}
    
    private void EventoMR()
    {
        string cadenaTemporal = string.Join("", conjuntoEventosM);
        cadenaRes = operacion.operation(cadenaTemporal);
    }
    
    private void EventoML()
    {
        if (EventosM(cadenaRes)) cadenaRes=operacion.Factorial(cadenaRes);
    }
    
    private void EventoXn(string apuntador)
    {
        if (EventosM(apuntador)) cadenaRes += "^";
    }
    
    private void EventoRaiz(string apuntador)
    {
         cadenaRes += "√";
    }
    
    private void EventoMSuma(string apuntador)
    {
        if (EventosM(apuntador))
        {
            string cadenaTemporal = conjuntoEventosM.Count == 0 ? operacion.operation(cadenaRes) : "+" + operacion.operation(cadenaRes);
            conjuntoEventosM.Add(cadenaTemporal);
        }
    }
    
    private void EventoMResta(string apuntador)
    {
        if (EventosM(apuntador)) conjuntoEventosM.Add("-" + operacion.operation(cadenaRes));
    }
    
    private void EventoIgual()
    {
        cadenaRes = operacion.operation(cadenaRes);
    }
    
    private void EventoSignos(string apuntador)
    {
        if (!string.IsNullOrEmpty(cadenaRes) && !"+-x/.^".Contains(apuntador)) cadenaRes += cadena;
    }
    
    private bool EventosM(string apuntador)
    {
        return !string.IsNullOrEmpty(cadenaRes) && !"+-x/.^".Contains(apuntador);
    }
}

